app/models/cinstance.rb

```diff
@@ -147,8 +147,8 @@ def validate_plan_is_unique?
     where(['cinstances.created_at <= ?', period.end])
   }
 
-  def self.provided_by(account)
-    joins(:service).references(:service).merge(Service.of_account(account)).readonly(false)
+  def self.provided_by(account, services_association: Service.unscoped)
+    where(plan_id: ApplicationPlan.unscoped.where(issuer_type: "Service", issuer_id: Service.unscoped.select(:id).of_account(account).merge(services_association)))
```

Comment:
In my tests, it seems to me the more efficient query is what @mayorova suggested in the issue [comment](https://issues.redhat.com/browse/THREESCALE-11139?focusedId=25054244&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-25054244)

```sql
SELECT
  COUNT(*)
FROM
  (
  SELECT
      DISTINCT `cinstances`.`id`
  FROM    `cinstances`
  INNER JOIN `services` ON
      `services`.`id` = `cinstances`.`service_id`
  WHERE
      `cinstances`.`type` = 'Cinstance'
      AND `services`.`account_id` = 2445579871111
      AND `services`.`state` != 'deleted') AS whatever;
```

And I can get a pretty similar query by using this code:

```suggestion
  def self.provided_by(account, services_association: Service.unscoped)
    joins(:service).where(services: { account_id: account.id }).merge(services_association)
```

Why make it more complicated adding a join to `plans`?

You mentioned [here](https://issues.redhat.com/browse/THREESCALE-11139?focusedId=25088853&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-25088853) that is better to get the services from the plans, but why?

Reply:
Yeah, but we spoke that getting through Plans is more reliable. I'm not sure why `service_id` exists in the model, maybe for service contracts. The original intention seems to be obtaining service through the plans.

This replicates that intent.

I'm fine if your investigation shows this would not be an issue though.

Reply:
Also, your changes generate a query which has an execution cost of `10.66`, while the suggested changes generate a query costing `3,23`.

However, I'm not sure how relevant is the execution cost for this case, since the original query had a cost of `2,72`... and took more than one second to execute. Do you know the cause of this contradiction between execution time and cost? It should be the less cost the faster...

CC @mayorova

Reply:
> Yeah, but we spoke that getting through Plans is more reliable. I'm not sure why service_id exists in the model, maybe for service contracts. The original intention seems to be obtaining service through the plans.

Understood, if the query is much faster now and we can be more confident it behaves as before, LGTM.

Reply:
@jlledom I have no idea, unfortunately, how the cost is calculated, and how it is related to the time of execution :shrug:

Reply:
@akostadinov I pulled the last changes from this branch and now this is the generated query:

```sql
SELECT
	COUNT(*)
FROM
	`cinstances`
WHERE
	`cinstances`.`type` = 'Cinstance'
	AND `cinstances`.`plan_id` IN (
	SELECT
		`plans`.`id`
	FROM
		`plans`
	WHERE
		`plans`.`type` = 'ApplicationPlan'
		AND `plans`.`issuer_type` = 'Service'
		AND `plans`.`issuer_id` IN (
		SELECT
			`services`.`id`
		FROM
			`services`
		WHERE
			`services`.`state` != 'deleted'
			AND `services`.`account_id` = 2445579871111))
```

Which is exactly the same as before the PR, but removing the spurious subquery. In my tests, there's not performance improvement, both old and new queries take the same time.

Reply:
Now I changed back to using a join with Services and just add to the services conditions instead of refering to plans and using sub-queries. Submitting to the old decision to keep service and plan->issuer in sync.